hybrid/compartment_examples/inter_comp_call_secure/main.c

```diff
@@ -0,0 +1,194 @@
+#error "This example only works on CHERI hybrid mode"
+#endif
+
+/*******************************************************************************
+ * Assembly Functions & Consts
+ ******************************************************************************/
+
+extern int switch_compartment();
+extern void comp_f_fn();
+extern void comp_g_fn();
+
+/*******************************************************************************
+ * Types & Consts
+ ******************************************************************************/
+
+const size_t comp_stack_size = 1992;
+const size_t total_comp_size = 5000;
+size_t id = 0;
+
+// Duplicate certain capabilities within the heap of each compartments;
+// currently we save DDC and PCC of `switch_compartment`
+const size_t saved_caps_count = 2;
+void *__capability saved_caps[saved_caps_count];
+
+/* Abstract representation of a compartment. Within 80 bytes we represent:
+ * - an id (8B)
+ * - address to the start of the compartment's stack (8B)
+ *   the size of the stack (8B)
+ * - address to the start of the compartment's heap (8B)
+ *   the size of the heap (8B)
+ * - alignment padding (8B)
+ * - the ddc corresponding to the compartment (16B)
+ * - the function within the compartment to be executed upon switching (16B)
+ */
+struct comp
+{
+	size_t id;
+	void *stack_addr;
+	size_t stack_len;
+	void *heap_addr;
+	size_t heap_len;
+	void *__capability ddc;
+	void *__capability comp_fn;
+};
+
+// ASM offsets, included here for validation
+#include "main.h"
+
+static_assert(COMP_SIZE == sizeof(struct comp), "Invalid `COMP_SIZE` provided");
+static_assert(COMP_OFFSET_STK_ADDR == offsetof(struct comp, stack_addr),
+			  "Invalid `COMP_OFFSET_STK_ADDR` provided.");
+static_assert(COMP_OFFSET_STK_LEN == offsetof(struct comp, stack_len),
+			  "Invalid `COMP_OFFSET_STK_LEN` provided.");
+static_assert(COMP_OFFSET_DDC == offsetof(struct comp, ddc), "Invalid `COMP_OFFSET_DDC` provided.");
+static_assert(COMP_OFFSET_PCC == offsetof(struct comp, comp_fn),
+			  "Invalid `COMP_OFFSET_PCC` provided.");
+
+struct comp comps[COMP_COUNT];
+
+/*******************************************************************************
+ * Privileged Functions
+ ******************************************************************************/
+
+void init_comps()
+{
+	void *__capability switch_cap = (void *__capability) switch_compartment;
+	switch_cap = cheri_bounds_set(switch_cap, 80 * 4);
+	asm("seal %w0, %w0, rb" : "+r"(switch_cap) :);
+	saved_caps[0] = switch_cap;
+
+	void *__capability comps_addr = (void *__capability) &comps;
+	comps_addr = cheri_bounds_set(comps_addr, COMP_COUNT * COMP_SIZE);
+	saved_caps[1] = comps_addr;
+}
+
+/* This function stores the compartment data in memory, so that it can be
+ * accessed by the compartments as needed.
+ */
+void executive_switch(struct comp c)
+{
+	void *__capability comps_addr = saved_caps[1];
+
+	asm("mov c19, %w0\n\t"
+		"mov x0, #0\n\t"
+		"msr CID_EL0, c0"
```

Comment:
I suspect that passing this in a general-purpose register might be better. It can be set in `switch_compartment`, rather than set here and read there.

The real answer here rather depends on what the ID actually means.

Reply:
Yes, honestly, `c19` is more a remnant of previous iterations. It can be anything, or if we can change the design of `switch_compartment` to not require this argument be passed somehow...

If by ID you mean the contents of `CID_EL0`, it represents the ID of the compartment we are jumping into (which would, of course, require some knowledge between compartments of what compartment corresponds to what ID). This currently corresponds to the index of the `comps` array.

Reply:
I meant `CID_EL0`. One of the first things that `switch_compartment` does is to read the value out of `CID_EL0` (into `c10`), to use it to index the `comps` array, but then why not just leave it in a general-purpose register (`c0`) and let `switch_compartment` choose if or how to put it in `CID_EL0`? That seems like a better abstraction to me, and a clearer interface because it looks more like a conventional function call.

Reply:
Indeed, there is no practical reason, beyond showing one way of using this "special-purpose-but-not-really" CID register. I can either add a comment to emphasise this, or we can use `c0` or any other register, if we prefer to have a "cleaner" example.

Reply:
I think it's good to set it, but we should do so in `switch_compartment`. Here, we should just pass the value in `c0` (or similar). I don't think using `CID_EL0` as a parameter-passing register would be considered a good idea.

Reply:
Removed the use of `CID_EL0` as discussed in https://github.com/capablevms/cheri-examples/pull/55/commits/47e39b76d2af704cb87d2bf7653a3b2c5b032d47.